packages/flutter/lib/src/material/action_buttons.dart

```diff
@@ -0,0 +1,427 @@
+/// `ModalRoute.of(context)?.canPop` to check whether the current route can be
+/// popped. If that value is false (e.g., because the current route is the
+/// initial route), the [BackButton] will not have any effect when pressed,
+/// which could frustrate the user.
+///
+/// Requires one of its ancestors to be a [Material] widget.
+///
+/// See also:
+///
+///  * [AppBar], which automatically uses a [BackButton] in its
+///    [AppBar.leading] slot when the [Scaffold] has no [Drawer] and the
+///    current [Route] is not the [Navigator]'s first route.
+///  * [BackButtonIcon], which is useful if you need to create a back button
+///    that responds differently to being pressed.
+///  * [IconButton], which is a more general widget for creating buttons with
+///    icons.
+///  * [CloseButton], an alternative which may be more appropriate for leaf
+///    node pages in the navigation tree.
+class BackButton extends _ActionButton {
+  /// Creates an [IconButton] with the appropriate "back" icon for the current
+  /// target platform.
+  const BackButton({
+    super.key,
+    super.color,
+    super.style,
+    super.onPressed,
+  }) : super(icon: const BackButtonIcon());
+
+  @override
+  void _onPressedCallback(BuildContext context) => Navigator.maybePop(context);
+
+  @override
+  String _getTooltip(BuildContext context) {
+    return MaterialLocalizations.of(context).backButtonTooltip;
+  }
+}
+
+/// A "close" icon that's appropriate for the current [TargetPlatform].
+///
+/// The current platform is determined by querying for the ambient [Theme].
+///
+/// See also:
+///
+///  * [CloseButton], an [IconButton] with a [CloseButtonIcon] that calls
+///    [Navigator.maybePop] to return to the previous route.
+///  * [IconButton], which is a more general widget for creating buttons
+///    with icons.
+///  * [Icon], a Material Design icon.
+///  * [ThemeData.platform], which specifies the current platform.
+class CloseButtonIcon extends StatelessWidget {
+  /// Creates an icon that shows the appropriate "close" image for
+  /// the current platform (as obtained from the [Theme]).
+  const CloseButtonIcon({ super.key });
+
+  @override
+  Widget build(BuildContext context) {
+    return _ActionIcon(
+      iconBuilderCallback: (ActionIconThemeData? actionIconTheme) {
+        return actionIconTheme?.closeButtonIconBuilder;
+      },
+      getIcon: (BuildContext context) => Icons.close,
+      getAndroidSemanticsLabel: (MaterialLocalizations materialLocalization) {
+        return materialLocalization.closeButtonTooltip;
+      },
+    );
+  }
+}
+
+/// A Material Design close icon button.
+///
+/// A [CloseButton] is an [IconButton] with a "close" icon. When pressed, the
+/// close button calls [Navigator.maybePop] to return to the previous route.
+///
+/// The [onPressed] callback can, for instance, be used to pop the platform's navigation stack
+/// via [SystemNavigator] instead of Flutter's [Navigator] in add-to-app
+/// situations.
+///
```

Comment:
This would be a good place to explain that the style's iconColor parameter overrides this class's color parameter for states where the foregroundColor resolves to non-null.

A test for this case (where both color and ButtonStyle.iconColor are specified) should be included as well.